build-tools/packages/build-cli/src/commands/generate/changelog.ts

```diff
@@ -84,24 +85,40 @@ export default class GenerateChangeLogCommand extends BaseCommand<
 	}
 
 	/**
-	 * Removes any additional metadata from all changesets and writes the output back to the source files.
+	 * Removes any custom metadata from all changesets and writes the output back to the source files. Changesets that
+	 * apply to no packages are assumed to be changesets for the release notes only and are thus deleted as part of
+	 * canonicalization, because such changesets cannot be made canonical.
+	 *
 	 * **Note that this is a lossy action!** The metadata is completely removed. Changesets are typically in source
 	 * control so changes can usually be reverted.
+	 *
+	 * @returns an array of paths to the canonicalized changesets. Deleted changesets will not be included.
 	 */
-	private async stripAdditionalMetadata(releaseGroupRootDir: string): Promise<void> {
+	private async canonicalizeChangesets(releaseGroupRootDir: string): Promise<string[]> {
 		const changesetDir = path.join(releaseGroupRootDir, DEFAULT_CHANGESET_PATH);
 		const changesets = await loadChangesets(changesetDir, this.logger);
+		const changesetPaths: string[] = [];
 
-		const toWrite: Promise<void>[] = [];
+		const toWriteOrDelete: Promise<void>[] = [];
 		for (const changeset of changesets) {
+			// changesets that apply to no packages should be omitted from the changelogs.
+			if (Object.keys(changeset.metadata).length === 0) {
+				this.info(`Removing changeset not included in changelog: ${changeset.sourceFile}`);
+				toWriteOrDelete.push(rm(changeset.sourceFile));
```

Comment:
After reading https://github.com/microsoft/FluidFramework/wiki/Changesets#changesets-that-apply-to-no-packages I still don't get why we have to delete these. If the third party tools we are using fail to support changesets which have an empty package list, that would be surprising: if this is the case, that section in the docs (or the comments here int he code) should explicitly state that tool "x" errors if the list of packages are empty, so we do special handling of them, or something like that.

Reply:
Ha, I went to repro the error, and it looks like https://github.com/microsoft/FluidFramework/pull/22420 updated [@changesets/cli](https://www.npmjs.com/package/@changesets/cli) and the new version just silently ignores changesets with no packages. So it's safe to not delete such changesets now.

Reply:
> If the third party tools we are using fail to support changesets which have an empty package list, that would be surprising

I'm frankly a little surprised they just silently ignore them now. In the typical use-case, a changeset that applies to no packages would usually indicate a mistake, and silently ignoring them is essentially data loss. But it's less code for us, so huzzah.